tests/queries/0_stateless/03008_s3_plain_rewritable.sql

```diff
@@ -0,0 +1,22 @@
+-- Tags: no-fasttest
+-- Tag: no-fasttest -- requires S3
+
+drop table if exists test_mt;
+create table test_mt (a Int32, b Int64) engine = MergeTree() order by a
+settings disk = disk(
+    type = s3_plain_rewritable,
+    endpoint = 'http://localhost:11111/test/test_mt/',
+    access_key_id = clickhouse,
+    secret_access_key = clickhouse);
+
+insert into test_mt (*) values (1, 2), (2, 2), (3, 1), (4, 7), (5, 10), (6, 12);
+insert into test_mt (*) select number, number from numbers_mt(1000000);
```

Comment:
It will also be good to check and validate content of `prefix.path` file, in integration test in will probably be more convenient.

Reply:
Also the layout https://pastila.nl/?000db959/a9e05708623112738049177aca209439#G5jPgWHeHVknjx4Tkz3L0g==
where we make random only part directory: `./store/df3/df3bc07c-0967-43f7-adbc-9daaaa8d629d/mlkufoyckwdnhxfo/` suggests that moving parts between tables will not work, e.g. for example `ALTER TABLE MOVE PARTITION TO TABLE`.
As I see https://github.com/ClickHouse/ClickHouse/issues/58347 suggests "Let the disk store each directory as a pseudorandom static prefix in S3", e.g. in my impression it means each directory should have a randomized name, while now it is not each. Seems there is no reason not to do it for each directory name, WDYT?

Reply:
Also may be it will be good to make a test where a table has a projection, because it will make files layout more complex (one more directory layer).

Reply:
Currently, the disk does not support `moveFile` required for `materialize projections`
Also, from the task description (projection is mutation):
> Let's also forbid ALTERs and mutations for the tables on this disk

Query `alter table test_mt materialize projection test_mt_projection`:

```
6. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: void std::__function::__policy_invoker<void (std::shared_ptr<DB::IMetadataTransaction>)>::__call_impl<std::__function::__default_alloc_func<DB::DiskObjectStorageTransaction::moveFile(String const&, String const&)::$_0, void (std::shared_ptr<DB::IMetadataTransaction>)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::IMetadataTransaction>&&) @ 0x000000000f3a0761
7. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::(anonymous namespace)::PureMetadataObjectStorageOperation::execute(std::shared_ptr<DB::IMetadataTransaction>) @ 0x000000000f39fe8f
8. ./build/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:891: DB::DiskObjectStorageTransaction::commit() @ 0x000000000f39f89b
9. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::DiskObjectStorage::moveFile(String const&, String const&, bool) @ 0x000000000f3905e8
10. ./contrib/llvm-project/libcxx/include/string:1499: DB::MergeTreeMutationEntry::commit(unsigned long) @ 0x0000000010ff158d
11. ./contrib/llvm-project/libcxx/include/string:1499: DB::StorageMergeTree::startMutation(DB::MutationCommands const&, std::shared_ptr<DB::Context const>) @ 0x0000000010fc86b2
12. ./build/./src/Storages/StorageMergeTree.cpp:0: DB::StorageMergeTree::mutate(DB::MutationCommands const&, std::shared_ptr<DB::Context const>) @ 0x0000000010fcb627
13. ./contrib/llvm-project/libcxx/include/vector:543: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000000fb36b50
14. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:71: DB::InterpreterAlterQuery::execute() @ 0x000000000fb3434c
15. ./build/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x00000000101db067
```
I can implement `moveFile` with the limitation that the filename does not change and see what happens.

Reply:
Not related to the PR, it looks like there is a bug in change setting that became visible if the disk is specified through `SETTINGS disk = disk(...)`in `CREATE TABLE`  and then the table is altered:
```
create table test_mt (a Int32, b Int64) engine = MergeTree() order by a
settings disk = disk(
    type = s3_plain_rewritable,
    endpoint = 'http://localhost:11111/test/test_mt/',
    access_key_id = clickhouse,
    secret_access_key = clickhouse);

alter table test_mt add column c Int64 after b;
```

```
2024.04.25 01:03:07.530107 [ 114749 ] {2a3cd2ac-1f4e-4e29-a720-a75ad7ca8e41} <Error> TCPHandler: Code: 170. DB::Exception: Bad get: has CustomType, requested String. (BAD_GET), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x00000000134b1232
1. ./build/./src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000afab639
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000063d0ecc
3. DB::Exception::Exception<std::basic_string_view<char, std::char_traits<char>>, DB::Field::Types::Which const&>(int, FormatStringHelperImpl<std::type_identity<std::basic_string_view<char, std::char_traits<char>>>::type, std::type_identity<DB::Field::Types::Which const&>::type>, std::basic_string_view<char, std::char_traits<char>>&&, DB::Field::Types::Which const&) @ 0x00000000063dbd2b
4. auto& DB::Field::safeGet<String const&>() @ 0x00000000066d96d1
5. ./contrib/llvm-project/libcxx/include/string:1499: DB::MergeTreeSettingsTraits::Accessor::instance()::$_0::operator()() const::'lambda127'(DB::MergeTreeSettingsTraits::Data&, DB::Field const&)::__invoke(DB::MergeTreeSettingsTraits::Data&, DB::Field const&) @ 0x0000000010e6f40c
6. ./src/Core/BaseSettings.h:224: DB::BaseSettings<DB::MergeTreeSettingsTraits>::set(std::basic_string_view<char, std::char_traits<char>>, DB::Field const&) @ 0x0000000010e792ba
7. ./src/Core/BaseSettings.h:319: DB::BaseSettings<DB::MergeTreeSettingsTraits>::applyChange(DB::SettingChange const&) @ 0x0000000010d180e8
8. ./contrib/llvm-project/libcxx/include/__iterator/wrap_iter.h:100: DB::MergeTreeData::checkColumnFilenamesForCollision(DB::StorageInMemoryMetadata const&, bool) const @ 0x0000000010cc05db
9. ./build/./src/Storages/MergeTree/MergeTreeData.cpp:3245: DB::MergeTreeData::checkAlterIsPossible(DB::AlterCommands const&, std::shared_ptr<DB::Context const>) const @ 0x0000000010cba24f
10. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:0: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000000fb366f6
11. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:71: DB::InterpreterAlterQuery::execute() @ 0x000000000fb3434c
```

The type of the "disk" setting is set as `String`, that's why it's a `BAD_GET` error.
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeSettings.h#L171

I'm not sure why it is trying to change the disk setting during `ALTER TABLE` when it is not even specified.

It works fine `storage_policy` setting.

Reply:
`ALTER TABLE MOVE PARTITION TO TABLE` requires hard link creation

``` sql
create table test_mt_dst (a Int32, b Int64) engine = MergeTree() partition by intDiv(a, 10000) order by tuple(a, b)
settings storage_policy = 's3_plain_rewritable';
alter table test_mt_src move partition 0 to table test_mt_dst;
```

```
4. ./src/Disks/ObjectStorages/IMetadataStorage.h:151: DB::IMetadataTransaction::throwNotImplemented() @ 0x000000000f3e2af9
5. ? @ 0x000000000f3e1c86
6. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: void std::__function::__policy_invoker<void (std::shared_ptr<DB::IMetadataTransaction>)>::__call_impl<std::__function::__default_alloc_func<DB::DiskObjectStorageTransaction::createHardLink(String const&, String const&)::$_0, void (std::shared_ptr<DB::IMetadataTransaction>)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::IMetadataTransaction>&&) @ 0x000000000f3980c1
7. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::(anonymous namespace)::PureMetadataObjectStorageOperation::execute(std::shared_ptr<DB::IMetadataTransaction>) @ 0x000000000f38efef
8. ./build/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:891: DB::DiskObjectStorageTransaction::commit() @ 0x000000000f38e9fb
9. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::DiskObjectStorage::createHardLink(String const&, String const&, bool) @ 0x000000000f38030f
10. ./build/./src/Storages/MergeTree/localBackup.cpp:0: DB::(anonymous namespace)::localBackupImpl(std::shared_ptr<DB::IDisk> const&, DB::IDiskTransaction*, String const&, String const&, DB::ReadSettings const&, DB::WriteSettings const&, bool, unsigned long, std::optional<unsigned long>, bool, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>> const&) @ 0x0000000010fa0685
11. ./build/./src/Storages/MergeTree/localBackup.cpp:0: DB::localBackup(std::shared_ptr<DB::IDisk> const&, String const&, String const&, DB::ReadSettings const&, DB::WriteSettings const&, bool, std::optional<unsigned long>, bool, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>> const&, std::shared_ptr<DB::IDiskTransaction>) @ 0x0000000010f9f8f8
12. ./contrib/llvm-project/libcxx/include/string:1499: DB::DataPartStorageOnDiskBase::freeze(String const&, String const&, DB::ReadSettings const&, DB::WriteSettings const&, std::function<void (std::shared_ptr<DB::IDisk> const&)>, DB::IDataPartStorage::ClonePartParams const&) const @ 0x0000000010bce704
13. ./build/./src/Storages/MergeTree/MergeTreeData.cpp:7098: DB::MergeTreeData::cloneAndLoadDataPartOnSameDisk(std::shared_ptr<DB::IMergeTreeDataPart const> const&, String const&, DB::MergeTreePartInfo const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::IDataPartStorage::ClonePartParams const&, DB::ReadSettings const&, DB::WriteSettings const&) @ 0x0000000010ce1cd6
14. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::StorageMergeTree::movePartitionToTable(std::shared_ptr<DB::IStorage> const&, std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>) @ 0x0000000010fce793
15. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::MergeTreeData::movePartitionToTable(DB::PartitionCommand const&, std::shared_ptr<DB::Context const>) @ 0x0000000010cc6851
16. ./build/./src/Storages/MergeTree/MergeTreeData.cpp:5163: DB::MergeTreeData::alterPartition(std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::vector<DB::PartitionCommand, std::allocator<DB::PartitionCommand>> const&, std::shared_ptr<DB::Context const>) @ 0x0000000010cc75b7
17. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:815: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000000fb25283
18. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:71: DB::InterpreterAlterQuery::execute() @ 0x000000000fb2290c
```

From the task: 
> Let's also forbid ALTERs and mutations for the tables on this disk, so we never do hard link operations.

Reply:
> Let the disk store each directory as a pseudorandom static prefix in S3", e.g. in my impression it means each directory should have a randomized name, while now it is not each. Seems there is no reason not to do it for each directory name, WDYT?

That was my initial implementation.
I changed it because `prefix.path` will be created for directories that are static (will never be be renamed or moved), such as

the top-level subdirectories `store`, `7d5` and `7d563340-6f73-463b-bc78-95baf347d19a` (table uuid) in the example `store/7d5/7d563340-6f73-463b-bc78-95baf347d19a/tmp_insert_all_2_2_0/`.

`CommonPathPrefixKeyGenerator` tokenizes the path by `/` and searches for the longest prefix starting from the last token.
If no prefix is found, the local path is used as-is. 
If there is a prefix, the next token after the match (a subdirectory) is mapped to a generated name.
Thus, `CommonPathPrefixKeyGenerator` supports more than one level of pseudo-random names, it is required, however, that `createDirectory` for the parent path to be called before it is called for the subdirectories on any level.

Reply:
> Currently, the disk does not support moveFile required for materialize projections
Also, from the task description (projection is mutation):

Ah, I did not know it needed `moveFile`. Could we explicitly disable projections then?

> Not related to the PR, it looks like there is a bug in change setting that became visible if the disk is specified through SETTINGS disk = disk(...)in CREATE TABLE and then the table is altered

Wow, thanks for catching. Created an issue https://github.com/ClickHouse/ClickHouse/issues/63019, I'll fix it (it was me who implemented this feature to specify disk via `SETTINGS disk = disk(...)`)

> ALTER TABLE MOVE PARTITION TO TABLE requires hard link creation

ah, ok, I thought it does `moveFile` instead, then yes - no need to support this

> I changed it because prefix.path will be created for directories that are static (will never be be renamed or moved), such as the top-level subdirectories store, 7d5 and 7d563340-6f73-463b-bc78-95baf347d19a (table uuid) in the example store/7d5/7d563340-6f73-463b-bc78-95baf347d19a/tmp_insert_all_2_2_0/.

Yeah, makes sense. Could you please add a comment about this in code? So that it will be clear for the code reader why we store some directories as is. 

btw may be we could get rid of `./store/` in object storage keys? or at least from `./` in the beginning? Getting rid of `store` could be good for s3 read performance, as it behaves slower when we read a lot of keys with the same prefix concurrently.

Reply:
> Could we explicitly disable projections then?

Explicitly disabled alters and mutations for plain type:
https://github.com/ClickHouse/ClickHouse/pull/61116/commits/1033ddfc9de311892ceb528dd0baebd96d53a985